File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        int ans = 0, curSum = 0, n = nums.size();
        unordered_map<int, int>mp;
        mp[0] = 1;
        for(int i=0; i<n; i++) {
            curSum += nums[i];
            int rem = (curSum%k + k)%k;
            if(mp.find(rem) != mp.end()) {
                ans += mp[rem];
            }
            mp[rem]++;
        }
        return ans;
    }
};

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        int curSum = 0, ans = 0, n = nums.length;
        HashMap<Integer, Integer>mp = new HashMap<>();
        mp.put(0, 1);
        for(int i=0; i<n; i++) {
            curSum += nums[i];
            int rem = (curSum%k + k)%k;
            if(mp.containsKey(rem)) {
                ans += mp.get(rem);
            }
            mp.put(rem, mp.getOrDefault(rem, 0) + 1);
        }
        return ans;
    }
}